plugins/org.eclipse.n4js/src/org/eclipse/n4js/workspace/N4JSWorkspaceConfig.java

```diff
@@ -247,27 +258,43 @@ private WorkspaceChanges detectAddedRemovedProjects(WorkspaceConfigSnapshot oldW
 		for (URI uri : Sets.union(oldProjectsMap.keySet(), newProjectsMap.keySet())) {
 			boolean isOld = oldProjectsMap.containsKey(uri);
 			boolean isNew = newProjectsMap.containsKey(uri);
+
 			if (isOld && !isNew) {
+				// deleted
 				changes = changes.merge(WorkspaceChanges.createProjectRemoved(oldProjectsMap.get(uri)));
 			} else if (!isOld && isNew) {
+				// added
 				ProjectConfigSnapshot newPC = configSnapshotFactory
 						.createProjectConfigSnapshot(newProjectsMap.get(uri));
 				changes = changes.merge(WorkspaceChanges.createProjectAdded(newPC));
 			} else if (isOld && isNew) {
-				if (alsoDetectChangedProjects) {
-					ProjectConfigSnapshot oldPC = oldProjectsMap.get(uri);
+				// check name change
+				String oldPackageName = ((N4JSProjectConfigSnapshot) oldProjectsMap.get(uri)).getPackageName();
+				String newPackageName = newProjectsMap.get(uri).getPackageName();
+				if (Objects.equals(oldPackageName, newPackageName)) {
+					// no change
+					if (alsoDetectChangedProjects) {
+						ProjectConfigSnapshot oldPC = oldProjectsMap.get(uri);
+						ProjectConfigSnapshot newPC = configSnapshotFactory
+								.createProjectConfigSnapshot(newProjectsMap.get(uri));
+
+						changes = changes.merge(N4JSProjectConfig.computeChanges(oldPC, newPC));
+					}
+				} else {
+					// names changed
 					ProjectConfigSnapshot newPC = configSnapshotFactory
 							.createProjectConfigSnapshot(newProjectsMap.get(uri));
-
-					changes = changes.merge(N4JSProjectConfig.computeChanges(oldPC, newPC));
+					changes = changes.merge(WorkspaceChanges.createProjectAdded(newPC));
+					changes = changes.merge(WorkspaceChanges.createProjectRemoved(oldProjectsMap.get(uri)));
```

Comment:
Ah! So behind this point, the old rule that project (now: package) names never change and instead projects are removed/added still holds? Great!

Reply:
Yes, true. The reason behind this is that when the package names changes, the dependency relations also might change and need to be recomputed.